### PR TITLE
[Feature](partitions) Support auto partition

### LINKED ISSUE #18547: [Feature] 根据分区列自动创建分区
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/doris/issues?q=is%3Aissue) and found no similar issues.

### Description

指定一个 DATE 或者 DATETIME 类型的分区列，不需要预先创建出大量分区，会在写入新数据时根据实际数据自动创建对应分区。

### Use case

1.不需要提前了解数据分区范围，根据数据value值自动创建分区
2.数据分区列如果出现没有提前建立出的分区，导入时不会出错
3.可节省数据同步时，去了解数据时间范围的工作

### Related issues

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zclllyybb**
## Proposed changes

Issue Number: close #18547 

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/exec/tablet_info.cpp [hunk 1]
```diff
@@ -213,6 +214,27 @@ VOlapTablePartitionParam::VOlapTablePartitionParam(std::shared_ptr<OlapTableSche
         _partition_block.insert(
                 {slot->get_empty_mutable_column(), slot->get_data_type_ptr(), slot->col_name()});
     }
+
+    if (t_param.__isset.enable_automatic_partition && t_param.enable_automatic_partition) {
+        _is_auto_partiton = true;
+        DCHECK(t_param.__isset.partition_function_exprs);
+        //TODO: Is there a way to avoid the dependency of slot index?
+        DCHECK(t_param.partition_function_exprs[0].nodes[0].node_type ==
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:222]: delete the useless dcheck

File: be/src/exec/tablet_info.cpp [hunk 2]
```diff
@@ -213,6 +214,27 @@ VOlapTablePartitionParam::VOlapTablePartitionParam(std::shared_ptr<OlapTableSche
...
+               TExprNodeType::FUNCTION_CALL)
+                << t_param.partition_function_exprs[0].nodes[0].node_type;
+        DCHECK(t_param.partition_function_exprs[0].nodes[1].node_type == TExprNodeType::SLOT_REF)
+                << t_param.partition_function_exprs[0].nodes[1].node_type;
+        DCHECK(t_param.partition_function_exprs[0].nodes[2].node_type ==
+               TExprNodeType::STRING_LITERAL)
+                << t_param.partition_function_exprs[0].nodes[2].node_type;
+
+        vectorized::VExpr::create_expr_tree(t_param.partition_function_exprs[0],
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:231]: `partition_function_exprs` seems a list？

File: be/src/exec/tablet_info.cpp [hunk 3]
```diff
@@ -213,6 +214,27 @@ VOlapTablePartitionParam::VOlapTablePartitionParam(std::shared_ptr<OlapTableSche
...
+                                            _part_func_ctx); //RETURN IF ERROR
+        _partition_function = _part_func_ctx->root();
+        _part_interval =
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:234]: if not the data_trun() function. maybe do not have child(1） please recheck the logic

File: be/src/exec/tablet_info.cpp (hunk 2 above)
Comment (zclllyybb) [reply be/src/exec/tablet_info.cpp:231]: yes. for multi partition keys&functions. now we only support one partition key column. so use [0]

File: be/src/exec/tablet_info.cpp (hunk 3 above)
Comment (zclllyybb) [reply be/src/exec/tablet_info.cpp:224]: will do in next pr. FE send the interval_slot_ref to BE.

Comment (HappenLee) [reply be/src/exec/tablet_info.cpp:223]: I think the work should do this PR， the way introduce lots of useless code

Comment (zclllyybb) [reply be/src/exec/tablet_info.cpp:223]: Yes. fixed

File: be/src/exec/tablet_info.cpp [hunk 4]
```diff
@@ -271,60 +283,14 @@ Status VOlapTablePartitionParam::init() {
         };
     }
 
-    DCHECK(!_t_param.partitions.empty()) << "must have at least 1 partition";
-    _is_in_partition = _t_param.partitions[0].__isset.in_keys;
+    // empty for auto partition table. checked in caller.
+    _is_in_partition = _t_param.partitions.empty() ? false : _t_param.partitions[0].__isset.in_keys;
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:287]: if in_partition empty(), the check way is not right

File: be/src/exec/tablet_info.cpp [hunk 5]
```diff
@@ -456,4 +476,54 @@ Status VOlapTablePartitionParam::_create_partition_key(const TExprNode& t_expr,
     return Status::OK();
 }
 
+Status VOlapTablePartitionParam::add_partitions(
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:479]: seems no where call this function?

**Review by HappenLee — COMMENTED** [review]
ADD Regression test for the feature

File: be/src/vec/sink/vtablet_sink.cpp [hunk 6]
```diff
@@ -364,6 +394,7 @@ Status VNodeChannel::init(RuntimeState* state) {
     _timeout_watch.start();
 
     // Initialize _cur_add_block_request
+    //FIXME: 把这些存一个param下来给requests用。
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:397]: do not use chinese

File: be/src/vec/runtime/vdatetime_value.cpp [hunk 7]
```diff
@@ -50,6 +50,29 @@ static const char* s_ab_month_name[] = {"",    "Jan", "Feb", "Mar", "Apr", "May"
 
 static const char* s_ab_day_name[] = {"Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun", nullptr};
 
+TimeUnit parse_upper_to_timeunit(const std::string& str) {
```
Comment (HappenLee) [inline be/src/vec/runtime/vdatetime_value.cpp:53]: remove the unless code

File: be/src/vec/runtime/vdatetime_value.h [hunk 8]
```diff
@@ -69,6 +71,8 @@ enum TimeUnit {
     YEAR_MONTH
 };
 
+TimeUnit parse_upper_to_timeunit(const std::string& str);
```
Comment (HappenLee) [inline be/src/vec/runtime/vdatetime_value.h:74]: remove

File: be/src/exec/tablet_info.cpp [hunk 9]
```diff
@@ -457,4 +475,54 @@ Status VOlapTablePartitionParam::_create_partition_key(const TExprNode& t_expr,
...
+                   << ", part_index=" << part->indexes[j].index_id
+                   << ", schema_index=" << _schema->indexes()[j]->index_id;
+                return Status::InternalError(
+                        "partition's index is not equal with schema's"
+                        ", part_index={}, schema_index={}",
+                        part->indexes[j].index_id, _schema->indexes()[j]->index_id);
+            }
+        }
+        _partitions.emplace_back(part);
+        for (auto& in_key : part->in_keys) {
```
Comment (HappenLee) [inline be/src/exec/tablet_info.cpp:520]: recheck the logic only in partition have `in_keys`, here must dispose the in partition and range partiton

File: be/src/vec/sink/vtablet_sink.cpp [hunk 10]
```diff
@@ -381,17 +409,20 @@ Status VNodeChannel::init(RuntimeState* state) {
     return Status::OK();
 }
 
-void VNodeChannel::open() {
+void VNodeChannel::_open_internal(int64_t index_id, std::vector<PTabletWithPartition>& tablets,
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:412]: where use the param `index_id` ?

File: be/src/vec/sink/vtablet_sink.cpp [hunk 11]
```diff
@@ -637,6 +594,16 @@ Status VNodeChannel::add_block(vectorized::Block* block, const Payload* payload,
     }
 
     SCOPED_RAW_TIMER(&_stat.append_node_channel_ns);
+    // Initialize _cur_add_block_request
+    //FIXME: move to init
+    auto cur_add_block_request = PTabletWriterAddBlockRequest {};
+    cur_add_block_request.set_allocated_id(&_parent->_load_id);
+    for (const auto& [index_id, tablets] : _index_tablets_map) {
+        cur_add_block_request.set_index_id(index_id);
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:602]: recheck the logic

File: be/src/vec/sink/vtablet_sink.cpp [hunk 12]
```diff
@@ -1328,30 +1538,114 @@ Status VOlapTableSink::send(RuntimeState* state, vectorized::Block* input_block,
...
+                }
+                const VOlapTablePartition* partition = nullptr;
+                bool is_continue = false;
+                uint32_t tablet_index = 0;
+                bool missing_this = false;
+                RETURN_IF_ERROR(_tablet_finder->find_tablet(state, block.get(), i, &partition,
+                                                            tablet_index, stop_processing,
+                                                            is_continue, &missing_this));
+                if (missing_this) {
+                    missing_vals_col.column->assume_mutable()->insert_from(*partition_col.column,
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:1568]: TODO:reduce the useless virtual function call

File: be/src/vec/sink/vtablet_sink.cpp [hunk 13]
```diff
@@ -1328,30 +1538,114 @@ Status VOlapTableSink::send(RuntimeState* state, vectorized::Block* input_block,
...
-                continue;
+
+            // for missing partition keys, calc the missing partition and save in _partitions_need_create
+            auto type = missing_vals_col.type;
+            if (missing_vals_col.column->size() > 0) {
+                vectorized::Block tmp_block = {missing_vals_col};
+                auto [part_ctx, part_func] = _get_partition_function();
+                int result_idx;
+                // calc the start value of missing partition ranges.
+                part_func->prepare(_state, *_output_row_desc, part_ctx.get());
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:1583]: the func only need to prepare one time, and must dispose the error status

File: be/src/vec/sink/vtablet_sink.cpp [hunk 14]
```diff
@@ -1328,30 +1538,114 @@ Status VOlapTableSink::send(RuntimeState* state, vectorized::Block* input_block,
...
+                // expose the data column
+                vectorized::ColumnPtr col = tmp_block.get_by_position(result_idx).column;
+                if (auto* nullable = check_and_get_column<vectorized::ColumnNullable>(*col)) {
+                    col = nullable->get_nested_column_ptr();
+                }
+                // calc the end value and save them.
+                if (type->get_type_id() == vectorized::TypeIndex::DateV2) {
+                    auto* arg_col = assert_cast<const vectorized::ColumnDateV2*>(col.get());
+                    for (int row = 0; row < arg_col->size(); row++) {
+                        _save_a_missing_val<vectorized::DateV2Value<vectorized::DateV2ValueType>>(
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:1597]: why not call type to string?

File: be/src/vec/sink/vtablet_sink.cpp [hunk 15]
```diff
@@ -1328,30 +1538,114 @@ Status VOlapTableSink::send(RuntimeState* state, vectorized::Block* input_block,
...
+                    }
+                } else if (type->get_type_id() == vectorized::TypeIndex::DateTime) {
+                    auto* arg_col = assert_cast<const vectorized::ColumnDateTime*>(col.get());
+                    for (int row = 0; row < arg_col->size(); row++) {
+                        _save_a_missing_val<vectorized::VecDateTimeValue>(
+                                arg_col, row, convert_buffer, PrimitiveType::TYPE_DATETIME);
+                    }
+                }
+                // then call FE to create it. then FragmentExecutor will redo the load.
+                _automatic_create_partition();
```
Comment (HappenLee) [inline be/src/vec/sink/vtablet_sink.cpp:1621]: here need to check the ` _automatic_create_partition` status

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 46.87 seconds
 stream load tsv:          536 seconds loaded 74807831229 Bytes, about 133 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.0 seconds inserted 10000000 Rows, about 344K ops/s
 storage size: 17162096371 Bytes

**Comment by zclllyybb** [comment]
run buildall

**Comment by hello-stephen** [comment]
(From new machine)TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 49.31 seconds
 stream load tsv:          541 seconds loaded 74807831229 Bytes, about 131 MB/s
 stream load json:         20 seconds loaded 2358488459 Bytes, about 112 MB/s
 stream load orc:          64 seconds loaded 1101869774 Bytes, about 16 MB/s
 stream load parquet:          31 seconds loaded 861443392 Bytes, about 26 MB/s
 insert into select:          29.4 seconds inserted 10000000 Rows, about 340K ops/s
 storage size: 17162166014 Bytes

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Comment by zclllyybb** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Pull request merged**